### PR TITLE
Check WAL file existence before upload

### PR DESCRIPTION
**Pull request by DustinChaloupka**
Issue: https://github.com/wal-g/wal-g/issues/104

A first pass attempt at not overwriting WAL files.

I did not see any way to set up unit tests for this. The `commands_test.go` file looked to just do parsing tests.

Manually building this and testing it did work though.

```
$ wal-g wal-push /var/lib/postgresql/10/main/pg_wal/000000010000000000000001
BUCKET: mybucket
SERVER: myserver
WAL PATH: myserver/wal_005/000000010000000000000001.lz4
$ wal-g wal-push /var/lib/postgresql/10/main/pg_wal/000000010000000000000001
BUCKET: mybucket
SERVER: myserver
2018/07/24 21:47:03 WAL file '/var/lib/postgresql/10/main/pg_wal/000000010000000000000001' already archived, not overwriting
```

### CODE REVIEW

File: commands.go [hunk 1]
```diff
@@ -593,6 +593,20 @@ func DownloadAndDecompressWALFile(pre *S3Prefix, walFileName string, dstLocation
...
+		Archive: aws.String(sanitizePath(tarUploader.server + WalPath + filepath.Base(dirArc) + "." + tarUploader.compressor.FileExtension())),
+	}
+
+	exists, err := archive.CheckExistence()
+	if err != nil {
+		log.Fatalf("%+v\n", err)
+	}
+	if exists {
+		log.Printf("WAL file '%s' already archived, not overwriting", dirArc)
+		return
```
Comment (rubbish) [inline commands.go]: Will this return a non-zero exit code? We'll want that so postgres knows the archive_command failed and will keep the WAL file around.

**Comment by x4m** [comment]
Is there a way to say to S3: "do not allow to overwrite this file"?

File: commands.go (hunk 1 above)
Comment (DustinChaloupka) [reply commands.go]: It exits with status `1` now instead.

**Comment by x4m** [comment]
Thanks for doing WAL-G better! I really appreciate your work on the project.

Things are a bit more complex.... WAL-G, invoked as archive_command can also get information from archive_status: it reduces startup cost and makes WAL archive parallel.
But in this PR you check the only file of archive_command, not those files which were stealed from archive_status by background uploader.

**Comment by DustinChaloupka** [comment]
@x4m The check has been moved to be included in the path of the `BgUploader` uploading WAL files as well.

**Comment by x4m** [comment]
Ugh...it breaks tests. If you have some spare time plz fix it. If you wish I can fix it myself tomorrow (UTC+5) and merge PR.

**Comment by x4m** [comment]
https://travis-ci.org/wal-g/wal-g/jobs/408531439

**Comment by DustinChaloupka** [comment]
@x4m I actually added a test for the `bg_uploader` to check that the overwrite does not happen as well.

**Comment by x4m** [comment]
`cmd := exec.Command(os.Args[0], "-test.run=TestBackgroundNoOverwriteWALUpload")`
whoa, hacky, tricky, but works.... testing panic... OK, I'll add comment there.

Thank you, pushed!

**Pull request merged**